libsplinter/src/error/constraint_violation.rs

```diff
@@ -0,0 +1,220 @@
+/// that column, an error with violation type `ConstraintViolationType::Unique` will occur.
+///
+/// Although this error maps closely to those generated by relational databases (such as those
+/// covered by Diesel), the underlying database does not need to be relational. It could, for
+/// example, be a memory or file-backed implementation of a store.
+pub struct ConstraintViolationError {
+    violation_type: ConstraintViolationType,
+    source: Option<Source>,
+}
+
+impl ConstraintViolationError {
+    /// Constructs a new `ConstraintViolationError` from a specified violation type.
+    ///
+    /// The implementation of `std::fmt::Display` for this error will use the
+    /// standard display of the ConstraintViolationType for it's message.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use splinter::error::{ ConstraintViolationError, ConstraintViolationType };
+    ///
+    /// let constraint_violation_error = ConstraintViolationError::with_violation_type(
+    ///     ConstraintViolationType::Unique
+    /// );
+    /// assert_eq!(format!("{}", constraint_violation_error), "ConstraintViolationError: Unique");
+    /// ```
+    pub fn with_violation_type(violation_type: ConstraintViolationType) -> Self {
+        Self {
+            violation_type,
+            source: None,
+        }
+    }
+
+    /// Constructs a new `ConstraintViolationError` from a specified source error and violation type.
+    ///
+    /// The implementation of `std::fmt::Display` for this error will simply pass through the
+    /// display of the source message unmodified.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use splinter::error::{ ConstraintViolationError, ConstraintViolationType };
+    ///
+    /// let db_err = std::io::Error::new(std::io::ErrorKind::Other, "db error");
+    /// let constraint_violation_error = ConstraintViolationError::from_source_with_violation_type(
+    ///     ConstraintViolationType::Unique,
+    ///     Box::new(db_err)
+    /// );
+    /// assert_eq!(format!("{}", constraint_violation_error), "db error: Unique");
+    /// ```
+    pub fn from_source_with_violation_type(
+        violation_type: ConstraintViolationType,
+        source: Box<dyn error::Error>,
+    ) -> Self {
+        Self {
+            violation_type,
+            source: Some(Source { source }),
+        }
+    }
+}
+
+impl error::Error for ConstraintViolationError {
+    fn source(&self) -> Option<&(dyn error::Error + 'static)> {
+        match &self.source {
+            Some(s) => Some(s.source.as_ref()),
+            None => None,
+        }
+    }
+}
+
+impl fmt::Display for ConstraintViolationError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        const TYPE_NAME: &str = "ConstraintViolationError";
+        match &self.source {
+            Some(s) => write!(f, "{}: {}", s.source, &self.violation_type),
+            None => write!(f, "{}: {}", TYPE_NAME, &self.violation_type),
```

Comment:
For the Some() case, just pass through source. Adding violation type output here likely provides no additional value.

For the None case, omit TYPE_NAME, as it unnecessarily exposes a coding detail to the user.